Add support of Keystone v3 'domain-name' to 'openstack-heat' cluster setup

micmro opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Keystone v3 authentication by user name [requires the domain (name or ID)](http://developer.openstack.org/api-ref/identity/v3/index.html?expanded=password-authentication-with-scoped-authorization-detail). If `domain-name` is not provided kubelet fails as seen below:
  
```
kubelet: error: failed to run Kubelet: could not init cloud provider "openstack": You must provide exactly one of DomainID or DomainName to authenticate by Username
systemd: kubelet.service: main process exited, code=exited, status=1/FAILURE
systemd: Unit kubelet.service entered failed state.
systemd: kubelet.service failed.
```

To solve this I pass a new`OS_USER_DOMAIN_NAME` environment variable through openstack-heat's heat-templates to write it as `domain-name` in `/srv/kubernetes/openstack.conf`. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #39783

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```
domain-name support for Keystone v3 added to openstack-heat cluster setup
```

Commit:
pass openstack-heat OS_USER_DOMAIN_NAME config to domain-name in openstack.conf

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39784)
<!-- Reviewable:end -->

zmerlynn commented:
@k8s-bot ok to test

@micmro: I have little context here. Do you know a reasonable reviewer?

micmro commented:
Perhaps one of the initial _openstack-heat_ provider committers like: @ohtake, @elsonrodriguez or @zreigz?

micmro commented:
Hi @ohtake, @elsonrodriguez or @zreigz would one of you mind checking this PR and give @zmerlynn a thumbs up or down?

zmerlynn commented:
Ping @ohtake @elsonrodriguez  @zreigz

elsonrodriguez commented:
I'm not in a position to test Openstack functionality, but it doesn't seem like there's a default for $OS_USER_DOMAIN_NAME. This may cause the deployment to error out. Is there a sane default for that value?

Commit:
add default for OS_USER_DOMAIN_NAME

micmro commented:
Thanks @elsonrodriguez - I've updated the PR and added a default. I could not think of a good default, as this is (literally 😄 ) domain specific, so I left it blank, which is would therefore result in a `/srv/kubernetes/openstack.conf` file like this (`$LBAAS_VERSION` is handled in the same way): 

```
...
tenant-name=ImATenant
domain-name=
[LoadBalancer]
...
```

micmro commented:
I am quite confident the failing of test `Jenkins GCE etcd3 e2e` is not related to this PR.

eparis commented:
@k8s-bot gce etcd3 e2e test this

eparis commented:
/approve

zmerlynn commented:
/lgtm
